MirServer/mir
# Disable the initial set of the CRTC when modes match

## How to test :test_tube: 
1. Provide the `run_with` command for your compositor with the new `SmoothBoothSupport` option. An example of adding this to `shell_main.cpp` would be: 

```cpp
...
return runner.run_with(
        {
            CursorTheme{"default:DMZ-White"},
            WaylandExtensions{},
            X11Support{},
            SmoothBootSupport{}, // Add me!
            window_managers,
            display_configuration_options,
            external_client_launcher,
            launcher,
            config_keymap,
            AppendEventFilter{quit_on_ctrl_alt_bksp},
            StartupInternalClient{spinner},
            ConfigurationOption{run_startup_apps, "startup-apps", "Colon separated list of startup apps", ""},
            pre_init(ConfigurationOption{[&](std::string const& typeface) { ::wallpaper::font_file(typeface); },
                                         "shell-wallpaper-font", "font file to use for wallpaper", ::wallpaper::font_file()}),
            ConfigurationOption{[&](std::string const& cmd) { terminal_cmd = cmd; },
                                "shell-terminal-emulator", "terminal emulator to use", terminal_cmd}
        });
```

2. Start your computer in a tty (this means that `gbm` cannot start. see [this link](systemctl set-default multi-user.target) on how to do that)
3. Once started, `ssh` into your computer from another computer as root
4. Run:
```sh
plymouthd
plymouth show-splash
plymouth deactivate
```
5. Then run your shell, e.g. `miral-shell`
6. Note that you should no longer see a black screen in-between plymouth and the compositor

## What's new :new: 
- Providing `create_display` with the options to that it can read whether or not it should do a smooth boot
- If it should do a smooth boot, the `Displaybuffer` will NOT initialize the initial visible frame buffer AND it will NOT set the CRTC for each output. This ensures that plymouth's previous contents remain on the screen

## Some caveats :hole: 
- We _may_ see a flicker if the next `set_crtc` changes the resolution or something

## Up next :next_track_button: 
I have another branch ready to go that will read the mapping into a Mir buffer and offer it back to the `Scene` to be displayed. However, this will call on us to transition the buffer out of existence, which is a bigger story
